### PR TITLE
Add the tool data_product_writer with unit test.

### PR DESCRIPTION
**Pull request by garthwatney**
| | |
|:---|:---|
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

Added the Data Product writer tool which reads a data product binary file and deserializes the data into a JSON output file.
The binary file may contain 1 or more Records which are specified in the FPP.  
A unit test is also added which reads in a collection of binary files and produces the resultant JSON data.  
The unit test includes a JSON Dictionary which is a manual translation of the `DpTool` component's FPP.   
By running the DpTool component's unit test will generate a collection of binary files.  `DpTool` and its unit test are in the F prime repo under `FpTools/dpTool`
In addition another JSON file is part of the tool which encapsulates the specification of the data product header information.

## Rationale

A tool to deserialize and display F prime data products.

## Testing/Review Recommendations

The `data_product_writer.py` is located in `src/fprime_gds/executables`  
The unit test is located in `test/fprime_gds/executables`  
Run `test_data_product_writer.py`.  The tool will be invoked for each *.bin file in this directory.  All the JSON output is placed in `JSONoutput`

## Future Work

Once the F prime JSON dictionary is part of the build, the tool will need to be changed to access this file.

### CODE REVIEW

**Comment by Joshua-Anderson** [comment]
Often times data products wind up storing opaque binary data - it might be nice to provide a mode that decodes the DP header/metadata to a json file but stores the contents of the dp as a binary data instead.

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py]: Removed redundant import struct

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py:71]: Removed the import from

File: test/fprime_gds/executables/test_data_product_writer.py
Comment (garthwatney) [reply test/fprime_gds/executables/test_data_product_writer.py]: sys is used at the end of handleException()

File: src/fprime_gds/executables/data_product_writer.py
Comment (thomas-bc) [inline src/fprime_gds/executables/data_product_writer.py]: Could we make this a `main()` function, and add

```
if __name__ == "main":
    main()
```

It will help with making it an entry-point for `fprime-cli ....`

File: src/fprime_gds/executables/plotData.py
Comment (thomas-bc) [inline src/fprime_gds/executables/plotData.py]: Also make this a `main()` function

File: src/fprime_gds/executables/data_product_writer.py
Comment (thomas-bc) [inline src/fprime_gds/executables/data_product_writer.py:1]: For this file to be executable through our utilities (e.g. `fprime-cli dp-writer dp.bin dict.json` or something like this), we will need to make it an [entrypoint](https://packaging.python.org/en/latest/specifications/entry-points/) to the tool.

Is this how this tool is intended to be used?

In that case, adding the entrypoint is rather simple, we just need a name (or subcommand name under `fprime-cli`) and to add it like we did for the other ones here: https://github.com/nasa/fprime-gds/blob/9c5f4f5f2d5a9ba1e80efb7094514e799b6383e0/pyproject.toml#L58-L63

File: test/fprime_gds/executables/dictionary.json
Comment (thomas-bc) [inline test/fprime_gds/executables/dictionary.json:1]: Can we move all the `.json` and `.bin` files under a subdirectory like `dp_writer_data/` or similar? If you prefer you can also move all files (including `test.py`) under `test/fprime_gds/executables/dp_writer`

File: src/fprime_gds/executables/data_product_writer.py
Comment (thomas-bc) [inline src/fprime_gds/executables/data_product_writer.py]: I'm not familiar with what this does, but does capitalization matter here? `EnumType` instead of `Enumtype` ?

File: test/fprime_gds/executables/test_data_product_writer.py
Comment (thomas-bc) [inline test/fprime_gds/executables/test_data_product_writer.py]: Instead of calling the script directly, we should import the source module (we do that [here](https://github.com/nasa/fprime-gds/blob/devel/test/fprime_gds/executables/test_run_deployment.py#L7)) and call its main function

File: test/fprime_gds/executables/test_data_product_writer.py
Comment (thomas-bc) [inline test/fprime_gds/executables/test_data_product_writer.py]: Instead of calling the script directly, we should import the source module and call its main function (we do that [here](https://github.com/nasa/fprime-gds/blob/devel/test/fprime_gds/executables/test_run_deployment.py#L7))

**Review by thomas-bc — CHANGES_REQUESTED** [review]
Since I was in there already, I made the few changes that make your test cases runnable in CI and with `pytest`, I hope you don't mind.

A few comments below:

File: src/fprime_gds/executables/data_product_writer.py
Comment (thomas-bc) [reply src/fprime_gds/executables/data_product_writer.py:1]: Maybe `fprime-dp-write`?

File: src/fprime_gds/executables/plotData.py
Comment (Joshua-Anderson) [inline src/fprime_gds/executables/plotData.py]: Does plotting data fall within the scope of what we want the fprime data product tools to do? There's lots of tools that exist for plotting time series data... I wonder if it might make more sense for the fprime data product tools to generate CSV files, allowing folks to plot their data in their favorite plotting tool.

Comment (thomas-bc) [reply src/fprime_gds/executables/plotData.py]: This code also seems specific to the `Ref/` app - maybe it would be a good example/guide to reference in the documentation instead?

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py]: Done

File: src/fprime_gds/executables/plotData.py
Comment (garthwatney) [reply src/fprime_gds/executables/plotData.py]: Done

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py:1]: I added fprime-dp-write as an entry point

File: test/fprime_gds/executables/dictionary.json
Comment (garthwatney) [reply test/fprime_gds/executables/dictionary.json:1]: I moved the dpspec.json to the dp_writer_data/
I didn't see any .bin files in that directory

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py]: That ForwardRef was not needed, I took it out

File: test/fprime_gds/executables/dictionary.json
Comment (garthwatney) [reply test/fprime_gds/executables/dictionary.json:1]: The json data that was in dpspec.json is moved into the python.
The dicitonary.json and all the binary files are in the test directory dp_writer_data

File: test/fprime_gds/executables/test_data_product_writer.py
Comment (garthwatney) [reply test/fprime_gds/executables/test_data_product_writer.py]: Fixed like all the other pytests

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py]: Fixed

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py]: Fixed

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py:71]: Fixed

File: test/fprime_gds/executables/test_data_product_writer.py
Comment (garthwatney) [reply test/fprime_gds/executables/test_data_product_writer.py]: sys is used

File: src/fprime_gds/executables/data_product_writer.py
Comment (thomas-bc) [inline src/fprime_gds/executables/data_product_writer.py:151]: Are those `header`, `headerHash` etc. types a specification of the DataProducts ? Or is this info enventually be read from the JSON Ground Dictionary (because user-specified in the FPP model)?
If this is going to be found in the dicitonary provided by the user, we should code that behavior in rather than using a stub here. And we should also wait for the FPP feature to be merged in before we can merge this PR.

File: src/fprime_gds/executables/data_product_writer.py
Comment (thomas-bc) [inline src/fprime_gds/executables/data_product_writer.py]: These global variables could probably be member variables of a DataProductWriter class that holds all the functions above.
Then `main()` could look like 
```
def main():
    args = parse_args()
    DataProductWriter(args.jsonDict, args.binFile).process()
```
 This would allow this code to be re-used as a library if we add more tools to process Data Products down the line. 
 If you have time to make this it would be nice, but we can defer to later.

**Review by thomas-bc — COMMENTED** [review]
The code looks good to me! Question about the dictionary, and one comment about reusability.

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py:151]: These are specifications of the header that hopefully will eventually be specified in the dictionary

File: src/fprime_gds/executables/data_product_writer.py
Comment (garthwatney) [reply src/fprime_gds/executables/data_product_writer.py]: I agree.  I removed the global variables and created a DataProductWriter class.

**Review by thomas-bc — DISMISSED** [review]
LGTM!

**Review by LeStarch — APPROVED** [review]

**Pull request merged**
